addons/project/models/project_task.py

```diff
@@ -301,6 +310,24 @@ def SELF_READABLE_FIELDS(self):
     def SELF_WRITABLE_FIELDS(self):
         return PROJECT_TASK_WRITABLE_FIELDS
 
+    def _inverse_project_id(self):
+        for task in self:
+            if task.project_id:
+                no_display_subtasks = task.child_ids.filtered(lambda t: not t.display_in_project)
+                if no_display_subtasks:
+                    no_display_subtasks.write({'project_id': task.project_id.id})
+
+    @api.depends('project_id')
+    def _compute_display_in_project(self):
+        self.filtered(
+            lambda t: not t.project_id or t.project_id != t.parent_id.project_id
+        ).display_in_project = True
+
+    @api.depends('project_id')
```

Comment:
Since we could also alter the `parent_id` field. 🙂 
```suggestion
    @api.depends('project_id', 'parent_id')
```